Jetstream Autoscaling Guide (#703)
GoogleCloudPlatform/ai-on-gke: tutorials-and-examples/inference-servers/jetstream/maxtext/single-host-inference/README.md

> ## Deploy Maxengine Server and HTTP Server

Reviewer: High level: why put the terraform files into the "tutorials and examples" folder rather than the benchmarks folder? The benchmarks folder is meant to have the terraform automation for the workloads, and it seems like we could organize this to have the "tutorials and examples" to refer to the terraform scripts in the benchmarking suite as needed.

Reviewer: discussed over chat, purpose is to have one off tutorial, not to add to benchmarking suite. Thus choices here relate more to tutorials-and-examples/ work. Refactoring tf files to happen in follow up PR.

kubectl apply -f deployment.yaml
```

### Deploy via Terraform

Navigate to the `./terraform` directory and do the standard [`terraform init`](https://developer.hashicorp.com/terraform/cli/commands/init). The deployment requires some inputs, an example `sample-terraform.tfvars` is provided as a starting point, run `cp sample-terraform.tfvars terraform.tfvars` and modify the resulting `terraform.tfvars` as needed. Finally run `terraform apply` to apply these resources to your cluster.

> #### (optional) Enable Horizontal Pod Autoscaling via Terraform

Reviewer 2: as discussed, using Terraform is good, but not all users would want to go with this approach. Please provide an approach to configure HPA with k8s yamls as another option

Author: Sure! I plan on immediately following up with the kubectl portion, separating them into two PRs since this one is already at 800+ lines.

> In situations where the deployment above is not sufficient for your inference load, you may consider metrics based horizontal pod autoscaling. To do this we recommend applying the following resources to your cluster:

Reviewer 2: ```suggestion
Applying the following resources to your cluster will enable autoscaling with customer metrics:
```

 - [Horizontal Pod Autoscaler (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/): For reading metrics and setting the maxengine-servers deployments replica count accordingly.

These components require a few more inputs and rerunning the [prior step](#deploy-via-terraform) with these set will deploy the components. The following input conditions should be satisfied: `custom_metrics_enabled` should be `true` and `metrics_port`, `hpa_type`, `hpa_averagevalue_target`, `hpa_min_replicas`, `hpa_max_replicas` should all be set.

 Note that only one HPA resource will be created. For those who want to scale based on multiple metrics, we recommend using the following template to apply more HPA resources:

```
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: jetstream-hpa
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: maxengine-server
  minReplicas: <YOUR_MIN_REPLICAS>
  maxReplicas: <YOUR_MAX_REPLICAS>
  metrics:
  - type: Pods
    pods:
      metric:
        name: prometheus.googleapis.com|<YOUR_METRIC_NAME>|gauge
      target:
        type: AverageValue
        averageValue: <YOUR_VALUE_HERE>

> ```

Reviewer 2: put this yaml in a file and reference it here

Reviewer 2: I'd suggest create a new folder called "autoscaling" under /single-host-inference, and in the autoscaling yaml, create a file 1. hpa.yaml (use prefill_backlog_size as the metric) 2. pod monitoring yaml 3. cmsa yaml

If you would like to probe the metrics manually, `cURL` your maxengine-server container on whatever metrics port you set and you should see something similar to the following:

> ```

Reviewer 2: where is the autoscaling with tpu hardware related part and scale rule?

Author: Added in [0dff592](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/703/commits/0dff5922c6d93983290bc0b08cf4ad2952065944)
